src/snowflake/snowpark/profiler.py

```diff
@@ -0,0 +1,207 @@
+        Args:
+            modules: List of names of stored procedures.
+        """
+        self.modules_to_register = modules
+        self._prepare_sql()
+        if self.session is not None:
+            self._register_modules()
+
+    def set_targeted_stage(self, stage: str):
+        """
+        Set targeted stage for profiler output.
+
+        Note:
+            The stage name must be a fully qualified name.
+
+        Args:
+            stage: String of fully qualified name of targeted stage
+        """
+        validate_object_name(stage)
+        self.stage = stage
+        self._prepare_sql()
+        if self.session is not None:
+            if (
+                len(self.session.sql(f"show stages like '{self.stage}'").collect()) == 0
+                and len(
+                    self.session.sql(
+                        f"show stages like '{self.stage.split('.')[-1]}'"
+                    ).collect()
+                )
+                == 0
+            ):
+                self.session.sql(
+                    f"create temp stage {self.stage} FILE_FORMAT = (RECORD_DELIMITER = NONE FIELD_DELIMITER = NONE )"
+                ).collect()
+            self._set_targeted_stage()
+
+    def set_active_profiler(self, active_profiler: str):
+        """
+        Set active profiler.
+
+        Note:
+            Active profiler must be either 'LINE' or 'MEMORY' (case-sensitive),
+            active profiler is set to 'LINE' by default.
+        Args:
+            active_profiler: String that represent active_profiler, must be either 'LINE' or 'MEMORY' (case-sensitive).
+
+        """
+        if active_profiler not in ["LINE", "MEMORY"]:
+            raise ValueError(
+                f"active_profiler expect 'LINE' or 'MEMORY', got {active_profiler} instead"
+            )
+        self.active_profiler = active_profiler
+        self._prepare_sql()
+        if self.session is not None:
+            self._set_active_profiler()
+
+    def _register_modules(self):
+        self.session.sql(self.register_modules_sql).collect()
+
+    def _set_targeted_stage(self):
+        self.session.sql(self.set_targeted_stage_sql).collect()
+
+    def _set_active_profiler(self):
+        self.session.sql(self.set_active_profiler_sql).collect()
+
+    def enable_profiler(self):
+        """
+        Enable profiler. Profiles will be generated until profiler is disabled.
+        """
+        self.session.sql(self.enable_profiler_sql).collect()
+
+    def disable_profiler(self):
+        """
+        Disable profiler.
+        """
+        self.session.sql(self.disable_profiler_sql).collect()
+
+    def _get_last_query_id(self):
+        pattern = r"WITH\s+.*?\s+AS\s+PROCEDURE\s+.*?\s+CALL\s+.*"
+        for query in self.query_history.queries[::-1]:
+            if query.sql_text.startswith("CALL") or re.match(
+                pattern, query.sql_text, re.DOTALL
+            ):
```

Comment:
Is `startswith` case-sensitive? If so we should uppercase all the strings before matching